### DISCUSSION TITLE
Compile time check for whether something is a @JsonClass in kotlin?

### DISCUSSION THREAD

**Discussion by vikasuy** [square/moshi#1585] [General]
Is there a way to leverage the type system to know whether a kotlin data class is a `@JsonClass` or not?

E.g. consider the following:

```
@JsonClass(generateAdapter = true)
data class Correct(foo: String)

// no @JsonClass definition
data class Incorrect(foo: String)

inline fun <reified T> serialize(t: T): String { 
  val adapter = moshi.adapter(T.class::java)
  return adapter.toJson(t)
}

serialize(Correct("foo")) // call succeeds
serialize(Incorrect("foo")) // fails at runtime but ideally fails at compile time
```

Is there a type bound I can use on `T` to restrict calling `serialize` on classes that have a `JsonClass` annotation (or at the very least, have a valid adapter defined)? I'm looking for a design that minimizes runtime errors.

**Comment by JakeWharton** [comment]
There is no way, no. An adapter for that type could have been registered. You are not required to use `@JsonClass` as a mechanism.
